02_knight_and_princess/kirill_tolstobrov/main.cpp

```diff
@@ -0,0 +1,181 @@
+      : id(init_id), vertex1_id(v1_init), vertex2_id(v2_init){};
+  operator std::string() const {
+    std::string result = "";
+    result += "{\n\"id\": " + std::to_string(id);
+    result += ",\n\"vertex_ids\": ";
+
+    result += "[" + std::to_string(vertex1_id);
+    result += ", " + std::to_string(vertex2_id);
+    result += "]\n}";
+
+    return result;
+  }
+
+  const EdgeId id;
+  const VertexId vertex1_id;
+  const VertexId vertex2_id;
+};
+
+class Graph {
+ public:
+  Graph() : vertex_id_counter(0), edge_id_counter(0) {}
+  int get_vertices_amount() const { return vertices_.size(); }
+  int get_edges_amount() const { return edges_.size(); }
+
+  VertexId get_new_vertex_id() { return vertex_id_counter++; }
+  EdgeId get_new_edge_id() { return edge_id_counter++; }
+
+  void add_new_vertex() { vertices_.emplace_back(get_new_vertex_id()); }
+
+  bool check_vertex_existence(const VertexId& vertex_id) const {
+    for (const auto& vertex : vertices_) {
+      if (vertex_id == vertex.id) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool are_vertices_connected(const VertexId& id1, const VertexId& id2) {
+    assert(check_vertex_existence(id1) &&
+           "Attemptig to access to nonexistent vertex: Error.");
+    assert(check_vertex_existence(id2) &&
+           "Attemptig to access to nonexistent vertex: Error.");
+    for (const auto& edge : connections_map_[id1]) {
+      if (edge->vertex1_id == id2 || edge->vertex2_id == id2) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void bind_vertices(const VertexId& id1, const VertexId& id2) {
+    assert(check_vertex_existence(id1) &&
+           "Attemptig to access to nonexistent vertex: Error.");
+    assert(check_vertex_existence(id2) &&
+           "Attemptig to access to nonexistent vertex: Error.");
+    assert(!are_vertices_connected(id1, id2) &&
+           "Attemptig to connect connected vertices: Error.");
+    const auto& edge = edges_.emplace_back(get_new_edge_id(), id1, id2);
+    connections_map_[id1].push_back(&edge);
+    connections_map_[id2].push_back(&edge);
+    vertices_[id1].add_edge(edge.id);
+    vertices_[id2].add_edge(edge.id);
+  }
+
+  operator std::string() const {
+    std::string result = "{\n  \"vertices\": [\n";
+
+    for (int i = 0; i < vertices_.size(); i++) {
+      result += std::string(vertices_[i]);
+      if (i != vertices_.size() - 1)
+        result += ",\n";
+    }
+
+    result += "\n  ],\n  \"edges\": [\n";
+
+    for (int i = 0; i < edges_.size(); i++) {
+      result += std::string(edges_[i]);
+      if (i != edges_.size() - 1)
+        result += ",\n";
+    }
+
+    result += "\n  ]\n}";
+    return result;
+  }
+
+ private:
+  std::vector<Edge> edges_;
+  std::vector<Vertex> vertices_;
+
+  VertexId vertex_id_counter;
+  EdgeId edge_id_counter;
+
+  // connections_map: vertex1 -> edge -> vertex2
+  std::map<VertexId, std::vector<const Edge*>> connections_map_;
```

Comment:
Договаривались не использовать сырые `С` поинтеры без крайней нужды.
```suggestion
  std::map<VertexId, std::vector<EdgeId>> connections_map_;
```